File: backend/app/services/movie_service.py
from app.integrations.tmdb_client import TMDbClient
from app.core.logger import logger 

tmdb_client = TMDbClient()

def get_discover_movies(
    page: int = 1,
    sort_by: str = "popularity.desc"
): 
    logger.info(f"Discover movies | page={page} | sort_by={sort_by}")

    data = tmdb_client.get(
        "/discover/movie", 
        params={
            "page": page,
            "sort_by": sort_by 
        }
    )

    return {
        "items": data["results"],
        "total": data["total_results"],
        "page": data["page"],
        "pageSize": len(data["results"]),
        "totalPages": data["total_pages"]
    }

def get_popular_movies(page: int = 1):
    logger.info(f"Fetching popular movies | page={page}")

    data = tmdb_client.get(
        "/movie/popular",
        params={
            "page": page
        }
    )

    return {
        "items": data["results"],
        "total": data["total_results"],
        "page": data["page"],
        "pageSize": len(data["results"]),
        "totalPages": data["total_pages"]
    }

def get_top_rated_movies(page: int = 1):
    logger.info(f"Fetching top rated movies | page={page}")

    data = tmdb_client.get(
        "/movie/top_rated",
        params={"page": page}
    )

    return {
        "items": data["results"],
        "total": data["total_results"],
        "page": data["page"],
        "pageSize": len(data["results"]),
        "totalPages": data["total_pages"]
    }

def get_trending_movies(
    page: int = 1,
    time_window: str = "week"
):
    logger.info(f"Fetching trending movies | page={page}")

    data = tmdb_client.get(
        f"/trending/movie/{time_window}",
        params={"page": page}
    )

    return {
        "items": data["results"],
        "total": data["total_results"],
        "page": data["page"],
        "pageSize": len(data["results"]),
        "totalPages": data["total_pages"]
    }

def get_movie_detail(movie_id: int):
    logger.info(
        f"Fetching movie detail | id={movie_id}"
    )

    movie = tmdb_client.get(
        f"/movie/{movie_id}",
        params={"append_to_response": "credits"}
    )

    return {
        **movie,
        "cast": movie.get("credits", {}).get("cast", []),
        "crew": movie.get("credits", {}).get("crew", [])
    }

def search_movies(
    query: str,
    page: int = 1
):
    logger.info(f"Searching movies | query={query}")

    data = tmdb_client.get(
        "/search/movie",
        params={
            "query": query,
            "page": page
        }
    )

    return {
        "items": data["results"],
        "total": data["total_results"],
        "page": data["page"],
        "pageSize": len(data["results"]),
        "totalPages": data["total_pages"]
    }

def get_genres():
    logger.info("Fetching movie genres")
    
    data = tmdb_client.get("/genre/movie/list")
    return data["genres"]

File: frontend/src/services/movieService.js
import api from "../utils/api";
import { API_ENDPOINTS } from "../utils/constants";

const handleError = (error, fallbackMessage) => {
  console.error(error);

  throw new Error(
    error.response?.data?.message || error.message || fallbackMessage,
  );
};

export const movieService = {
  async getMovies(page = 1, sort_by = "popularity.desc") {
    try {
      const response = await api.get(API_ENDPOINTS.GET_MOVIES, {
        params: { page, sort_by },
      });

      return response.data;
    } catch (error) {
      handleError(error, "Failed to fetch movies");
    }
  },

  async getTrending(page = 1, timeWindow = "week") {
    try {
      const response = await api.get(API_ENDPOINTS.GET_TRENDING, {
        params: { page, timeWindow },
      });

      return response.data;
    } catch (error) {
      handleError(error, "Failed to fetch trending movies");
    }
  },

  async getPopular(page = 1) {
    try {
      const response = await api.get(API_ENDPOINTS.GET_POPULAR, {
        params: { page },
      });

      return response.data;
    } catch (error) {
      handleError(error, "Failed to fetch popular movies");
    }
  },

  async getTopRated(page = 1) {
    try {
      const response = await api.get(API_ENDPOINTS.GET_TOP_RATED, {
        params: { page },
      });

      return response.data;
    } catch (error) {
      handleError(error, "Failed to fetch top rated movies");
    }
  },

  async getMovieDetail(id) {
    try {
      const response = await api.get(API_ENDPOINTS.GET_MOVIE_DETAIL(id));
      return response.data;
    } catch (error) {
      handleError(error, "Failed to fetch movie detail");
    }
  },

  async searchMovies(q, page = 1) {
    try {
      const response = await api.get(API_ENDPOINTS.SEARCH_MOVIES, {
        params: { q, page },
      });

      return response.data;
    } catch (error) {
      handleError(error, "Failed to search movies");
    }
  },

  async getGenres() {
    try {
      const response = await api.get(API_ENDPOINTS.GET_GENRES);

      return response.data;
    } catch (error) {
      handleError(error, "Failed to fetch genres");
    }
  },
};
